edx_arch_experiments/scripts/get_action_errors.py

```diff
@@ -1,138 +1,262 @@
+                # expected rate-limiting headers.
+                print("Rate-limited without timing hint; performing exponential backoff.")
+                backoff_s = 2 * backoff_s if backoff_s else 4
+                continue
+            else:
+                # Generic error case
+                response.raise_for_status()
+
+    def _write_json(self, data, *path_parts):
+        """
+        Write data as pretty-printed JSON to the given path (joining as needed).
+        """
+        with open(path.join(*path_parts), 'w') as f:
+            json.dump(data, f, sort_keys=True, indent=2)
+
+    def _log_attempt(self, attempt, workflow_dir):
+        """
+        Log this attempt and all of its checks and annotations.
+        """
+        attempt_dir = _ensure_dir(workflow_dir, f"attempt_{attempt['run_attempt']}")
+
+        attempt_file = path.join(attempt_dir, f"attempt.json")
+        if path.isfile(attempt_file):
+            print("Attempt already fully downloaded; skipping.")
+            return
+
+        # Get the checks associated with this workflow run -- this
+        # includes output title, summary, and text.
+        for check_run in self._github_get(attempt['check_suite_url'] + '/check-runs').json()['check_runs']:
+            self._write_json(check_run, attempt_dir, f"check_run_{check_run['id']}.json")
+
+            annotations = self._github_get(check_run['output']['annotations_url']).json()
+            self._write_json(annotations, attempt_dir, f"annotations_{check_run['id']}.json")
+
+        # Do this last, indicating that the attempt was completely
+        # downloaded. This allows us to skip it next time.
+        self._write_json(attempt, attempt_file)
+
+    def _list_all_attempts(self, run):
+        """
+        Yield all attempts of the given workflow run, including that one.
+        """
+        yield run
+        while next_url := run.get('previous_attempt_url'):
+            resp = self._github_get(next_url, params=self.workflow_fetch_params)
+            run = resp.json()
+            yield run
+
+    def _download_workflow_run(self, run, repo_dir):
+        print(f"Downloading workflow run id={run['id']}")
+
+        workflow_dir = _ensure_dir(repo_dir, f"run_{run['id']}",)
+
+        download_marker = path.join(workflow_dir, f"download-marker.json")
+        if path.isfile(download_marker):
+            print("Workflow already fully downloaded; skipping.")
+            return
+
+        # We're getting the *most recent attempt* of a run. Spool
+        # out the whole list of attempts and write them out.
+        for attempt in self._list_all_attempts(run):
+            self._log_attempt(attempt, workflow_dir)
+
+        # Once all attempts have been logged, write out a marker file
+        # that indicates this workflow has been completely downloaded
+        # and can be skipped in the future.
+        self._write_json(self.download_marker_data, download_marker)
```

Comment:
Author's note: Consider writing the latest attempt count here, allowing us to download newer attempts of workflows from previous executions of the script. Somewhat of an edge case.